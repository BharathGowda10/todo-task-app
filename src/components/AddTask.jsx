import PropTypes from "prop-types";
import { useForm } from "react-hook-form";
import { MdWarning } from "react-icons/md";

const AddTask = ({ onAdd }) => {
  const {
    handleSubmit,
    register,
    setValue,
    formState: { errors },
    reset,
    watch,
  } = useForm({
    defaultValues: {
      name: "",
      dateTime: "",
      type: "",
      priority: "",
    },
    mode: "onSubmit",
  });

  const selectedPriority = watch("priority");

  const onSubmit = (data) => {
    if (data.name.trim()) {
      onAdd(data);
      reset();
    }
  };

  const handlePriorityClick = (priority) => {
    setValue("priority", priority, { shouldValidate: true });
  };

  return (
    <form
      className="row mb-3 py-4 justify-content-center align-items-center gap-2 gap-md-0"
      onSubmit={handleSubmit(onSubmit)}
    >
      <div
        className="col-10 d-flex flex-column flex-md-row align-items-center justify-content-center"
        style={{ gap: "10px" }}
      >
        <div className="w-100">
          <input
            type="text"
            name="name"
            id="task-name"
            {...register("name", {
              required: "Input Field Is Required",
            })}
            className="form-control form-control-lg"
            placeholder="Enter a new task"
            autoComplete="off"
          />
          {errors?.name?.message && (
            <span className="text-danger ms-2 mt-1 fs-6 d-flex align-items-center">
              <MdWarning className="me-1 fs-5" />
              {errors.name.message}
            </span>
          )}
        </div>
        <button
          type="submit"
          id="add-task-button"
          className="btn submit-button btn-warning btn-lg mt-2 mt-md-0 align-self-start"
        >
          Add Task
        </button>
      </div>

      <div className="row mb-3">
        <div className="col-10 col-md-4 my-3 mb-md-0 mx-auto">
          <div className="border border-2 rounded p-3 text-center">
            <label htmlFor="task-datetime" className="form-label mb-2">
              Select Date &amp; Time
            </label>
            <input
              type="datetime-local"
              id="task-datetime"
              name="dateTime"
              {...register("dateTime", {
                required: "Date & Time Is Required",
              })}
              className="form-control"
              style={{ maxWidth: "320px", margin: "0 auto" }}
            />
            {errors?.dateTime?.message && (
              <span
                style={{ maxWidth: "320px" }}
                className="text-danger mx-auto mt-1 fs-6 d-flex align-items-center"
              >
                <MdWarning className="me-1 fs-5" />
                {errors.dateTime.message}
              </span>
            )}
          </div>
        </div>
        <div className="col-10 col-md-4 my-3 mb-md-0 mx-auto">
          <div className="border border-2 rounded p-3 text-center">
            <label htmlFor="task-type" className="form-label mb-2">
              Task Type
            </label>
            <select
              id="task-type"
              name="type"
              {...register("type", {
                required: "Select Any One Type",
              })}
              className="form-select"
              style={{ maxWidth: "320px", margin: "0 auto" }}
            >
              <option value="">Select type</option>
              <option value="Work">Work</option>
              <option value="Shopping">Shopping</option>
              <option value="Chores">Chores</option>
              <option value="Others">Others</option>
            </select>
            {errors?.type?.message && (
              <span
                style={{ maxWidth: "320px" }}
                className="text-danger mx-auto mt-1 fs-6 d-flex align-items-center"
              >
                <MdWarning className="me-1 fs-5" />
                {errors.type.message}
              </span>
            )}
          </div>
        </div>
        <div className="col-10 col-md-4 my-3 mx-auto">
          <div className="border border-2 rounded p-3 text-center">
            <div className="mb-2">
              <label className="form-label mb-2 d-block">Task Priority</label>
              <input
                type="hidden"
                {...register("priority", {
                  required: "Select Any One Priority",
                })}
              />
              <span
                className={`badge bg-danger fs-6 me-2 ${
                  selectedPriority === "High"
                    ? "border border-2 border-primary"
                    : ""
                }`}
                style={{ cursor: "pointer", border: "none" }}
                onClick={() => handlePriorityClick("High")}
              >
                High
              </span>
              <span
                className={`badge bg-warning text-dark fs-6 me-2 ${
                  selectedPriority === "Medium"
                    ? "border border-2 border-primary"
                    : ""
                }`}
                style={{ cursor: "pointer", border: "none" }}
                onClick={() => handlePriorityClick("Medium")}
              >
                Medium
              </span>
              <span
                className={`badge bg-success fs-6 ${
                  selectedPriority === "Low"
                    ? "border border-2 border-primary"
                    : ""
                }`}
                style={{ cursor: "pointer", border: "none" }}
                onClick={() => handlePriorityClick("Low")}
              >
                Low
              </span>
              {errors?.priority?.message && (
                <span
                  className="text-danger mt-1 fs-6 d-flex align-items-center justify-content-center"
                  style={{ width: "100%" }}
                >
                  <MdWarning className="me-1 fs-5" />
                  {errors.priority.message}
                </span>
              )}
            </div>
          </div>
        </div>
      </div>
    </form>
  );
};

AddTask.propTypes = {
  onAdd: PropTypes.func.isRequired,
};

export default AddTask;
