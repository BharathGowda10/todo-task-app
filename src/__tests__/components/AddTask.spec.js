import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import AddTask from '../../components/AddTask';

describe('AddTask Component', () => {
  it('match snap shot', () => {
    const { asFragment } = render(<AddTask onAdd={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders Input and Button', () => {
    render(<AddTask onAdd={jest.fn()} />);
    expect(
      screen.getByPlaceholderText(/enter a new task/i)
    ).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: /add task/i })
    ).toBeInTheDocument();
  });
  it('should validate for empty input box', async () => {
    const onAdd = jest.fn();
    render(<AddTask onAdd={onAdd} />);
    const button = screen.getByRole('button', { name: /add task/i });
    await userEvent.click(button);
    expect(
      await screen.findByText(/input field is required/i)
    ).toBeInTheDocument();
    expect(
      await screen.findByText(/Date & Time Is Required/i)
    ).toBeInTheDocument();
    expect(await screen.findByText(/select any one type/i)).toBeInTheDocument();
    expect(
      await screen.findByText(/select any one priority/i)
    ).toBeInTheDocument();
    expect(onAdd).not.toHaveBeenCalled();
  });

  it('submit form successfully', async () => {
    const onAdd = jest.fn();
    render(<AddTask onAdd={onAdd} />);
    const input = screen.getByPlaceholderText(/enter a new task/i);
    await userEvent.type(input, 'Task 1');
    const dateTimeInput = screen.getByLabelText(/select date & time/i);
    userEvent.clear(dateTimeInput);
    fireEvent.change(dateTimeInput, { target: { value: '2023-10-01T10:00' } });
    const typeSelect = screen.getByLabelText(/task type/i);
    await userEvent.selectOptions(typeSelect, 'Work');
    const highPriorityBadge = screen.getByText(/high/i);
    userEvent.click(highPriorityBadge);
    const button = screen.getByRole('button', { name: /add task/i });
    userEvent.click(button);
    await waitFor(() =>
      expect(onAdd).toHaveBeenCalledWith({
        name: 'Task 1',
        dateTime: '2023-10-01T10:00',
        type: 'Work',
        priority: 'High',
      })
    );
  });
  it('should highlight the selected priority badge with border-primary', async () => {
    render(<AddTask onAdd={jest.fn()} />);
    const highBadge = screen.getByText(/high/i);
    const mediumBadge = screen.getByText(/medium/i);
    const lowBadge = screen.getByText(/low/i);

    // Initially, none should have border-primary
    expect(highBadge).not.toHaveClass('border-primary');
    expect(mediumBadge).not.toHaveClass('border-primary');
    expect(lowBadge).not.toHaveClass('border-primary');

    // Click High
    await userEvent.click(highBadge);
    expect(highBadge).toHaveClass('border-primary');
    expect(mediumBadge).not.toHaveClass('border-primary');
    expect(lowBadge).not.toHaveClass('border-primary');

    // Click Medium
    await userEvent.click(mediumBadge);
    expect(mediumBadge).toHaveClass('border-primary');
    expect(highBadge).not.toHaveClass('border-primary');
    expect(lowBadge).not.toHaveClass('border-primary');

    // Click Low
    await userEvent.click(lowBadge);
    expect(lowBadge).toHaveClass('border-primary');
    expect(highBadge).not.toHaveClass('border-primary');
    expect(mediumBadge).not.toHaveClass('border-primary');
  });
});
