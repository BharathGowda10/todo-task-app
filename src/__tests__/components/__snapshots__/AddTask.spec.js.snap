// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTask Component match snap shot 1`] = `
<DocumentFragment>
  <form
    class="row mb-3 py-4 justify-content-center align-items-center gap-2 gap-md-0"
  >
    <div
      class="col-10 d-flex flex-column flex-md-row align-items-center justify-content-center"
      style="gap: 10px;"
    >
      <div
        class="w-100"
      >
        <input
          autocomplete="off"
          class="form-control form-control-lg"
          id="task-name"
          name="name"
          placeholder="Enter a new task"
          type="text"
        />
      </div>
      <button
        class="btn submit-button btn-warning btn-lg mt-2 mt-md-0 align-self-start"
        id="add-task-button"
        type="submit"
      >
        Add Task
      </button>
    </div>
    <div
      class="row mb-3"
    >
      <div
        class="col-10 col-md-4 my-3 mb-md-0 mx-auto"
      >
        <div
          class="border border-2 rounded p-3 text-center"
        >
          <label
            class="form-label mb-2"
            for="task-datetime"
          >
            Select Date & Time
          </label>
          <input
            class="form-control"
            id="task-datetime"
            name="dateTime"
            style="max-width: 320px; margin: 0px auto;"
            type="datetime-local"
          />
        </div>
      </div>
      <div
        class="col-10 col-md-4 my-3 mb-md-0 mx-auto"
      >
        <div
          class="border border-2 rounded p-3 text-center"
        >
          <label
            class="form-label mb-2"
            for="task-type"
          >
            Task Type
          </label>
          <select
            class="form-select"
            id="task-type"
            name="type"
            style="max-width: 320px; margin: 0px auto;"
          >
            <option
              value=""
            >
              Select type
            </option>
            <option
              value="Work"
            >
              Work
            </option>
            <option
              value="Shopping"
            >
              Shopping
            </option>
            <option
              value="Chores"
            >
              Chores
            </option>
            <option
              value="Others"
            >
              Others
            </option>
          </select>
        </div>
      </div>
      <div
        class="col-10 col-md-4 my-3 mx-auto"
      >
        <div
          class="border border-2 rounded p-3 text-center"
        >
          <div
            class="mb-2"
          >
            <label
              class="form-label mb-2 d-block"
            >
              Task Priority
            </label>
            <input
              name="priority"
              type="hidden"
              value=""
            />
            <span
              class="badge bg-danger fs-6 me-2 "
              style="cursor: pointer;"
            >
              High
            </span>
            <span
              class="badge bg-warning text-dark fs-6 me-2 "
              style="cursor: pointer;"
            >
              Medium
            </span>
            <span
              class="badge bg-success fs-6 "
              style="cursor: pointer;"
            >
              Low
            </span>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
